### PR TITLE
[SBOM] Replace BadgerDB cache by a custom LRU cache with a cleaning mechanism for Trivy

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?

Replaces BadgerDB cache by a new custom LRU cache for Trivy with a local storage backed by BoltDB and a clean up mechanism to remove unused entries. It has mechanisms to limit the size in memory and on disk.
The clean up mechanism required to store ArtifactIDs and BlobIDs in the workloadmeta.SBOM so that they can be removed from the cache.

It adds config options to parametrize the cache:
- `sbom.use_custom_cache`: Enables the custom cache.
- `sbom.custom_cache_max_disk_size`: Maximum total size of the cached objects. It is not equivalent to disk usage. If BoltDB uses extra storage, it will not limited by this config option.
- `sbom.custom_cache_max_cache_entries`: Maximum number of cached objects.
- `sbom.cache_clean_interval` : Duration interval of garbage collection.

It also adds the following metrics to monitor the custom cache:

- `sbom_cache_disk_size`: Total size, in bytes, of the database as reported by BoltDB.
- `sbom_cached_object_size`: Total size, in bytes, of cached SBOM objects on disk. Limited by `sbom.custom_cache_max_disk_size`
- `sbom_cache_hits_total`: Total number of cache hits
- `sbom_cache_misses_total`: Total number of cache misses
- `sbom_cache_evicts_total`: Total number of cache evicts

### Motivation

The default Trivy cache based on Boltdb doesn’t work for us, because it doesn’t have any cleanup mechanism. It is ok for a CLI based approach, but for a long running process like the Datadog-Agent it can trigger an high memory and disc usage.

We tried to implement a custom cache based on badgerDB that provide a TTL mechanism, but after some tests we discovered that the DB doesn't fit other criterias that we have:

BadgerDB doesn’t offer an easy way to update TTL of a specific entry

BadgerDB is optimized for large dataset (millions of entries). But it is not our case.

BadgerDB allocate a large disk space event if it doesn’t use it.

### Additional Notes
Tested and deployed on 7.44.x first: https://github.com/DataDog/datadog-agent/pull/16438

### Possible Drawbacks / Trade-offs
N/A. No disk or memory pressure observed during tests.

### Describe how to test/QA your changes

See https://github.com/DataDog/datadog-agent/pull/16438
Verify that the unit tests are passing.
Deploy the agent on an environment with `containerd` and/or `docker` with these settings:

```yaml
DD_TELEMETRY_ENABLED=TRUE
DD_CONTAINER_IMAGE_COLLECTION_METADATA_ENABLED=true
DD_SBOM_ENABLED=true
DD_SBOM_USE_CUSTOM_CACHE=true
DD_SBOM_CACHE_CLEAN_INTERVAL=180
```

- Pull an image. 

- Check the SBOM generation time with `agent workload-list --verbose`. In my case, the `datadog/agent:latest` images takes a few seconds.

- Now delete the image, pull it again, and check the generation time again. It should be faster because of the cache.

- Delete the image again and wait for the garbage collector which should run every 180s with the settings above.

- Pull the image again, it should take a few seconds.

- To test the max disk size and max cache entries, the unit tests should be enough. Otherwise you can add log lines to MissingBlobs, set low values to the following parameters and pull many images. 
```
DD_CONTAINER_IMAGE_COLLECTION_SBOM_CUSTOM_CACHE_MAX_CACHE_ENTRIES and DD_CONTAINER_IMAGE_COLLECTION_SBOM_CUSTOM_CACHE_MAX_DISK_SIZE 
```
* Be careful, if the values are too low, GetArtifact and GetBlob will return an error, making the scan fail. 

You should be able to see the metrics showing up:
`kube exec pod -- curl -s localhost:6000/telemetry | grep sbom`

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/util/trivy/containerd.go [hunk 1]
```diff
@@ -148,13 +148,17 @@ func inspect(ctx context.Context, imgMeta *workloadmeta.ContainerImageMetadata,
 	for k := range imgConfig.Config.ExposedPorts {
 		portSet[nat.Port(k)] = struct{}{}
 	}
+	created := ""
+	if lastHistory.Created != nil {
```
Comment (AliDatadog) [inline pkg/util/trivy/containerd.go:152]: For reviewers, it is possible that this field is nil, triggering a panic

File: pkg/sbom/scanner/scanner.go [hunk 2]
```diff
@@ -78,17 +78,21 @@ func (s *Scanner) start(ctx context.Context) {
 	if s.running {
 		return
 	}
-
+	cleanTicker := time.NewTicker(config.Datadog.GetDuration("sbom.cache_clean_interval"))
```
Comment (clamoriniere) [inline pkg/sbom/scanner/scanner.go:81]:
```suggestion
	go func() {
		cleanTicker := time.NewTicker(config.Datadog.GetDuration("sbom.cache_clean_interval"))
		defer cleanTicker.Stop()
```

File: pkg/sbom/scanner/scanner.go [hunk 3]
```diff
@@ -78,17 +78,23 @@ func (s *Scanner) start(ctx context.Context) {
 	if s.running {
 		return
 	}
-
+	go func() {
+		cleanTicker := time.NewTicker(config.Datadog.GetDuration("sbom.cache_clean_interval"))
+		defer cleanTicker.Stop()
 	go func() {
```
Comment (clamoriniere) [inline pkg/sbom/scanner/scanner.go:84]:
```suggestion
	go func() {
		cleanTicker := time.NewTicker(config.Datadog.GetDuration("sbom.cache_clean_interval"))
		defer cleanTicker.Stop()
```

File: pkg/util/trivy/trivy.go [hunk 4]
```diff
@@ -162,23 +161,34 @@ func NewCollector(cfg config.Config) (*Collector, error) {
...
-	fanalCache, err := config.CacheProvider()
+	fanalCache, cacheCleaner, err := config.CacheProvider()
 	if err != nil {
 		return nil, err
 	}
 
+	cleanTicker := time.NewTicker(cfg.GetDuration("sbom.cache_clean_interval"))
+	go func() {
+		// Cache can not be cleaned concurrently with a scan
+		for range cleanTicker.C {
```
Comment (clamoriniere) [inline pkg/util/trivy/trivy.go:172]:
```suggestion
	go func() {
	    cleanTicker := time.NewTicker(cfg.GetDuration("sbom.cache_clean_interval"))
	    defer cleanTicker.Close()
		// Cache can not be cleaned concurrently with a scan
		for range cleanTicker.C {
```

**Review by clamoriniere — APPROVED** [review]

**Pull request merged**